File: server/src/utils/clip-characters.js
/**
 * 跨 clip 角色在场回填（与 worker/utils/clip_characters.py 对齐）
 */

const ENTER_NEAR_NAME =
  /(?:进入|到达|走进|推门|进门|闯入|现身|赶来|到来|开门进入|推门进入|走进来|闯进来|推门而入|推门走进|出现在|来到)/;

const ENTER_GLOBAL =
  /(?:有人)?(?:进入|走进|推门进入|闯入|赶到|来到)(?:了)?(?:该|此|这个)?(?:房间|卧室|室内|场景|空间)/;

function clipText(clip) {
  return `${clip.content || ''}\n${clip.summary || ''}`;
}

export function characterEntersInClip(characterName, clip) {
  if (!characterName) return false;
  const text = clipText(clip);
  if (!text.trim()) return false;
  if (ENTER_GLOBAL.test(text)) return true;
  let m;
  const re = new RegExp(ENTER_NEAR_NAME.source, 'g');
  while ((m = re.exec(text)) !== null) {
    const start = Math.max(0, m.index - 40);
    const end = Math.min(text.length, m.index + m[0].length + 40);
    if (text.slice(start, end).includes(characterName)) return true;
  }
  return false;
}

function sameLocation(a, b) {
  const locA = (a.location || '').trim();
  const locB = (b.location || '').trim();
  if (!locA || !locB) return true;
  if (locA === locB) return true;
  const baseA = locA.split('_')[0];
  const baseB = locB.split('_')[0];
  return baseA === baseB;
}

/**
 * @param {Array<Record<string, unknown>>} clips
 * @returns {Array<Record<string, unknown>>}
 */
export function backfillClipCharacters(clips) {
  if (!clips?.length) return clips || [];

  const ordered = [...clips].sort(
    (a, b) => (a.clipIndex ?? 0) - (b.clipIndex ?? 0),
  );

  for (let i = 0; i < ordered.length - 1; i += 1) {
    const curr = ordered[i];
    const nxt = ordered[i + 1];
    if (!sameLocation(curr, nxt)) continue;

    const currList = [...(curr.characters || [])];
    const currSet = new Set(currList);
    for (const name of nxt.characters || []) {
      if (!name || currSet.has(name)) continue;
      if (characterEntersInClip(name, nxt)) continue;
      currList.push(name);
      currSet.add(name);
    }
    curr.characters = currList;
  }

  return ordered;
}

File: worker/utils/clip_characters.py
"""
跨 clip 角色在场回填：落实 analyze_story 提示中的「相邻 clip 连贯性回查」。

LLM 常把环境建立镜头（如清晨卧室光线变化）的 characters 只填主角，
遗漏同床已睡着的配角；下一段醒来/互动时才列出两人。本模块向前回填。
"""

from __future__ import annotations

import re
from typing import Any

# 下一段若出现这些词且与角色名邻近，视为「本段才进入场景」，不回填到上一段
_ENTER_NEAR_NAME = re.compile(
    r'(?:进入|到达|走进|推门|进门|闯入|现身|赶来|到来|开门进入|推门进入|'
    r'走进来|闯进来|推门而入|推门走进|出现在|来到)'
)

# 全段级进入（未点名也可判定本段为「入场」镜头）
_ENTER_GLOBAL = re.compile(
    r'(?:有人)?(?:进入|走进|推门进入|闯入|赶到|来到)(?:了)?(?:该|此|这个)?(?:房间|卧室|室内|场景|空间)'
)


def _clip_text(clip: dict[str, Any]) -> str:
    return f"{clip.get('content') or ''}\n{clip.get('summary') or ''}"


def character_enters_in_clip(character_name: str, clip: dict[str, Any]) -> bool:
    """该角色是否在本 clip 正文中被描写为「进入/到达」场景。"""
    if not character_name:
        return False
    text = _clip_text(clip)
    if not text.strip():
        return False
    if _ENTER_GLOBAL.search(text):
        return True
    for m in _ENTER_NEAR_NAME.finditer(text):
        start = max(0, m.start() - 40)
        end = min(len(text), m.end() + 40)
        window = text[start:end]
        if character_name in window:
            return True
    return False


def _same_location(a: dict[str, Any], b: dict[str, Any]) -> bool:
    loc_a = (a.get('location') or '').strip()
    loc_b = (b.get('location') or '').strip()
    if not loc_a or not loc_b:
        return True
    if loc_a == loc_b:
        return True
    # 「卧室_清晨」与「卧室」等同场
    base_a = loc_a.split('_')[0]
    base_b = loc_b.split('_')[0]
    return base_a == base_b


def backfill_clip_characters(clips: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """
    对相邻 clip 做向前回填：下一段在场且本段无「进入」描写的角色，补入上一段 characters。
    原地修改并返回按 clipIndex 排序的列表。
    """
    if not clips:
        return clips

    ordered = sorted(clips, key=lambda c: c.get('clipIndex', 0))
    for i in range(len(ordered) - 1):
        curr = ordered[i]
        nxt = ordered[i + 1]
        if not _same_location(curr, nxt):
            continue

        curr_list = list(curr.get('characters') or [])
        curr_set = set(curr_list)
        for name in nxt.get('characters') or []:
            if not name or name in curr_set:
                continue
            if character_enters_in_clip(name, nxt):
                continue
            curr_list.append(name)
            curr_set.add(name)
        curr['characters'] = curr_list

    return ordered
